lib/web/assistant.go

```diff
@@ -324,6 +325,16 @@ func runAssistant(h *Handler, w http.ResponseWriter, r *http.Request,
 			return trace.Wrap(err)
 		}
 
+		// We can not know how many tokens we will consume in advance.
+		// Try to consume a small amount of tokens first.
+		const lookaheadTokens = 100
+		if !h.assistantLimiter.AllowN(time.Now(), lookaheadTokens) {
+			if err := sendRateLimitedMessage(h, conversationID, ws); err != nil {
+				return trace.Wrap(err)
+			}
+			continue
```

Comment:
This should not break the WS connection. In theory, the user can remain on the conversation page, come back to it later and continue once the tokens are replenished.